src/uu/tail/src/follow/files.rs

```diff
@@ -0,0 +1,211 @@
+//  * This file is part of the uutils coreutils package.
+//  *
+//  * For the full copyright and license information, please view the LICENSE
+//  * file that was distributed with this source code.
+
+// spell-checker:ignore tailable seekable stdlib (stdlib)
+
+use crate::args::Settings;
+use crate::chunks::BytesChunkBuffer;
+use crate::paths::{HeaderPrinter, PathExtTail};
+use crate::text;
+use std::collections::hash_map::Keys;
+use std::collections::HashMap;
+use std::fs::{File, Metadata};
+use std::io::{stdout, BufRead, BufReader, BufWriter};
+
+use std::path::{Path, PathBuf};
+use uucore::error::UResult;
+
+/// Data structure to keep a handle on files to follow.
+/// `last` always holds the path/key of the last file that was printed from.
+/// The keys of the HashMap can point to an existing file path (normal case),
+/// or stdin ("-"), or to a non existing path (--retry).
+/// For existing files, all keys in the HashMap are absolute Paths.
+pub struct FileHandling {
+    map: HashMap<PathBuf, PathData>,
+    last: Option<PathBuf>,
+    header_printer: HeaderPrinter,
+}
+
+impl FileHandling {
+    pub fn from(settings: &Settings) -> Self {
+        Self {
+            map: HashMap::with_capacity(settings.inputs.len()),
+            last: None,
+            header_printer: HeaderPrinter::new(settings.verbose, false),
+        }
+    }
+
+    /// Wrapper for HashMap::insert using Path::canonicalize
+    pub fn insert(&mut self, k: &Path, v: PathData, update_last: bool) {
+        let k = Self::canonicalize_path(k);
+        if update_last {
+            self.last = Some(k.to_owned());
+        }
+        let _ = self.map.insert(k, v);
+    }
+
+    /// Wrapper for HashMap::remove using Path::canonicalize
+    pub fn remove(&mut self, k: &Path) -> PathData {
+        self.map.remove(&Self::canonicalize_path(k)).unwrap()
```

Comment:
Why is `unwrap` okay for these methods?

Reply:
I actually don't think it is and tend to return an Option or Result from these methods, but I need to dig deeper through the watch module to figure this out. These functions are heavily used. I can address this issue together with refactoring `watch` in the next pr.